WordPress/gutenberg
# Switch performance tests to Playwright

## What?

A follow-up to #51084

Closes #51383 
Closes #51379

This PR effectively switches our performance tests to be handled by Playwright. Apart from that, the following improvements have been made:

- Make sure all the tests are consistent in terms of methodology and that the outlying samples (as shown [here](https://github.com/WordPress/gutenberg/pull/51084#issuecomment-1607658262)) are filtered out to improve metrics stability.
- Use the [attach API](https://playwright.dev/docs/api/class-testinfo#test-info-attachments) to handle result files (as suggested by @kevin940726 [here](https://github.com/WordPress/gutenberg/pull/51084#discussion_r1231995414)).
- Make the [performance.js](https://github.com/WordPress/gutenberg/blob/a045f9ec5d8ec9efd186f484ca3fc4d050abb44c/bin/plugin/commands/performance.js) comparison runner result-agnostic (as suggested by @youknowriad [here](https://github.com/WordPress/gutenberg/pull/49238#issuecomment-1551500521)), and move the results to be handled by the dedicated [PerformanceReporter](https://github.com/WordPress/gutenberg/blob/1f8ecd1323cd238eb98447ec56888f5760a29811/test/performance/config/performance-reporter.ts) script.
- Make the PerformanceReporter handle results when running both locally and in CI. Previously, the Reporter was handling only local results which resulted in a discrepancy between CI and local results.
- Make the [performance.js](https://github.com/WordPress/gutenberg/blob/a045f9ec5d8ec9efd186f484ca3fc4d050abb44c/bin/plugin/commands/performance.js) runner suite-agnostic by automatically loading suites from the filesystem.
- Upload the curated results JSON to the artifacts zip as well for easier inspection.

ℹ️ **The [codevitals](https://www.codevitals.run/project/gutenberg) app requires no update as the test results will be produced and uploaded as expected.**

I ran a couple of comparisons of this branch vs. the base branch (round [1](https://github.com/WordPress/gutenberg/pull/52022/commits/d649aacefeb62d343d36a4d6b36a3fa701af5124), [2](https://github.com/WordPress/gutenberg/pull/52022/commits/ae40542df6325a04bfeb24ffe4015d447cc30434) and [3](https://github.com/WordPress/gutenberg/pull/52022/commits/18fe132a027b87d4025f52599ecb3465cad1617a)). The results seem stable and should not disrupt the current codevitals logs:

   | round 1 | round 2 | round 3 |
   | ------------- | ------------- | ------------- |
   | ![](https://github.com/WordPress/gutenberg/assets/1451471/2e0d5ba9-531e-4061-919c-1523accdee40) | ![](https://github.com/WordPress/gutenberg/assets/1451471/ea3e0b53-e5fa-4bd8-bf08-58490688db40) | ![](https://github.com/WordPress/gutenberg/assets/1451471/5b2ff08b-e5b4-4151-8866-c128cabad054) |


## Why?
As a part of the [Playwright migration task](https://github.com/WordPress/gutenberg/issues/38851), but also to improve the overall experience and quality of our performance metrics.


## Testing Instructions

- To run performance tests locally, the command remains unchanged, e.g.:

   ```
   npm run test:performance -- post-editor --headed
   ```
   
   You should see different reporter output though, as it was updated to be consistent with the format we use in CI. Also, missing front-end perf results have been added:
   
   | Before | After |
   | ------------- | ------------- |
   | ![](https://user-images.githubusercontent.com/1451471/245774797-5b0b4377-29ad-47fa-bcfa-735f5645e899.png) | ![](https://github.com/WordPress/gutenberg/assets/1451471/62f5c54d-20e8-4892-9315-58eb4f410069) |

- To run branch performance comparison tests locally (via the CLI tool), the command also remains unchanged as well as the output, e.g.:

   ```
   ./bin/plugin/cli.js perf my-branch trunk --tests-branch my-tests-branch
   ```
   
   You can check the performance GH action output for this PR to verify the above as well.
